### ISSUE TITLE
Host Integrations Require Chef 13/Ruby-2.4.0

### ISSUE THREAD

**Issue by mikeantonelli** [newrelic/infrastructure-agent-chef#14]
The metadata file constrains this cookbook to chef [12.14 or higher](https://github.com/newrelic/infrastructure-agent-chef/blob/0.1.0/metadata.rb), but the host_integrations recipe appears to [require an API](https://github.com/newrelic/infrastructure-agent-chef/blob/0.1.0/recipes/host_integrations.rb#L19) only available via chef-13.0+.

In chef 12.x the embedded ruby version only goes as high as 2.3.1 where [`Hash#compact` is not available](https://docs.ruby-lang.org/en/2.3.0/Hash.html). Chef 13.0 [upgraded to ruby 2.4.0]((https://docs.chef.io/release_notes.html#ruby-version-upgraded-to-2-4-1) ) where [`Hash#compact` is available](https://docs.ruby-lang.org/en/2.4.0/Hash.html#method-i-compact).

I'm using the following configuration:

```ruby
  newrelic_infra: {
    config: {
      log_file: '/var/log/newrelic_infra.log'
    },
    features: {
      host_integrations: true
    },
    host_integrations: {
      config: {
        mysql: {
          integration_name: 'com.newrelic.mysql',
          instances: [
            {
              command: 'status',
              arguments: {
                hostname: 'localhost',
                port: '3306',
                username: '...',
                password: '...',
                extended_metrics: 1,
                extended_innodb_metrics: 1
              },
              labels: {
                role: 'write-replica'
              }
            }
          ]
        }
      }
    }
  },
```

Output:
```
Recipe: newrelic-infra::host_integrations
  * yum_package[newrelic-infra-integrations] action install (up to date)
  * file[/etc/newrelic-infra/integrations.d/mysql.yaml] action create

    ================================================================================
    Error executing action `create` on resource 'file[/etc/newrelic-infra/integrations.d/mysql.yaml]'
    ================================================================================

    NoMethodError
    -------------
    undefined method `compact' for #<Hash:0x0000000c2c1238>

    Cookbook Trace:
    ---------------
    /var/chef/cache/cookbooks/newrelic-infra/recipes/host_integrations.rb:19:in `block (3 levels) in from_file'
    /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:78:in `run_action'
    /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:106:in `block (2 levels) in converge'
    /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:106:in `each'
    /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:106:in `block in converge'
    /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:105:in `converge'

    Resource Declaration:
    ---------------------
    # In /var/chef/cache/cookbooks/newrelic-infra/recipes/host_integrations.rb

     18:   file ::File.join(node['newrelic_infra']['host_integrations']['config_dir'], "#{integration_name}.yaml") do
     19:     content(lazy { YAML.dump(config.to_hash.compact.deep_stringify) })
     20:     owner node['newrelic_infra']['user']['name']
     21:     group node['newrelic_infra']['group']['name']
     22:     mode  '0640'
     23:     notifies :restart, 'poise_service[newrelic-infra]'
     24:   end
     25: end

    Compiled Resource:
    ------------------
    # Declared in /var/chef/cache/cookbooks/newrelic-infra/recipes/host_integrations.rb:18:in `block in from_file'

    file("/etc/newrelic-infra/integrations.d/mysql.yaml") do
      action [:create]
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      declared_type :file
      cookbook_name "newrelic-infra"
      recipe_name "host_integrations"
      content #<Chef::DelayedEvaluator:0x000000083eaaf0@/var/chef/cache/cookbooks/newrelic-infra/recipes/host_integrations.rb:19>
      owner "newrelic_infra"
      group "newrelic_infra"
      mode "0640"
      path "/etc/newrelic-infra/integrations.d/mysql.yaml"
    end

    Platform:
    ---------
    x86_64-linux
```

**Comment by taharah** [comment]
Oh wow we've been using 12.21.x on our nodes, which I guess has 2.4 bundled since it's been working

**Comment by mikeantonelli** [comment]
I just upgraded to `12.21.14` and saw the same issue. The chef [release notes for 13.0/13.1](https://docs.chef.io/release_notes.html#ruby-version-upgraded-to-2-4-1) cite an upgrade to `ruby-2.4`, I'm not sure how you got this working on `12.21.x`.

**Comment by smith** [comment]
@mikeantonelli Thanks for reporting this. We should be able to get this updated to be backward compatible.

**Comment by smith** [comment]
@mikeantonelli @taharah Please give #16 a try and let me know if it fixes it. If I get your reviews I should be able to merge on Monday morning.

**Comment by smith** [comment]
Fixed in #19.
